README.md
M-Pesa Ruby Gem
===============

[![Gem Version](https://badge.fury.io/rb/mpesa.svg)](https://badge.fury.io/rb/mpesa) [![Build Status](https://travis-ci.org/itsmrwave/mpesa-gem.svg?branch=master)](https://travis-ci.org/itsmrwave/mpesa-gem) [![Coverage Status](https://coveralls.io/repos/github/itsmrwave/mpesa-gem/badge.svg)](https://coveralls.io/github/itsmrwave/mpesa-gem)

Basically it's a gem that makes it easy to integrate your app with the new
[M-PESA platform][mpesa-api] dubbed G2 (for M-PESA 2nd generation platform). It
Handles all the SOAP stuff abstracting any direct interaction with the API
endpoints so that you can focus on what matters. _Building awesome_.

The gem [should be up on RubyGems.org][rubygems-mpesa], it's [accompanying API
reference here][rubydoc-mpesa], the [CHANGELOG here][changelog] and [all the
releases here][releases].


Installation
------------

Add this line to your application's Gemfile:

    gem 'mpesa'

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install mpesa


Issues
------

[Bug reports][issues] and [pull requests][pull-requests] are welcome on GitHub
at [github.com/itsmrwave/mpesa-gem][repositories]. This project is intended to
be a safe, welcoming space for collaboration, and [contributors][contributors]
are expected to adhere to the [Contributor Code of Conduct][code-of-conduct].


Development
-----------

After checking out the repo, run this to set up your local machine:

    $ bin/setup

To get an interactive prompt that will allow you to experiment:

    $ bin/console

For learning how to read a WSDL document, read the [Beginner's Guide by Thomas
Bayer](http://predic8.com/wsdl-reading.htm). It's a good idea to know what
you're working with and this might really help you debug certain problems.

The contributing flow/process I'd recommend [can be found here][contributing].


Documentation
-------------

You may start up a local documentation server at `http://0.0.0.0:8808` to
preview the gem API documentation locally. Just call:

    $ bundle exec yard server --reload

The result should be similar as to what will eventually be available as [the
RubyDoc API reference][rubydoc-mpesa].


Testing
-------

To run tests, call:

    $ bundle exec rspec spec

To check if the style guide has been violated:

    $ bundle exec rubocop


License
-------

[Job King'ori Maina](http://kingori.co/) © 2016. The MIT License bundled therein
is a permissive license that is short and to the point. It lets people do
anything they want as long as they provide attribution and waive liability. You
can [learn more about it here][mit-license].

[changelog]: https://raw.githubusercontent.com/itsmrwave/mpesa-gem/master/CHANGELOG.md
[code-of-conduct]: https://raw.githubusercontent.com/itsmrwave/mpesa-gem/master/CODE_OF_CONDUCT.md
[contributing]: https://raw.githubusercontent.com/itsmrwave/mpesa-gem/master/CONTRIBUTING.md
[contributors]: https://github.com/itsmrwave/mpesa-gem/graphs/contributors
[issue-new]: https://github.com/itsmrwave/mpesa-gem/issues/new
[issues]: https://github.com/itsmrwave/mpesa-gem/issues
[mit-license]: http://choosealicense.com/licenses/mit
[mpesa-api]: http://www.safaricom.co.ke/business/corporate/m-pesa-payments-services/m-pesa-api
[pull-requests]: https://github.com/itsmrwave/mpesa-gem/pulls
[releases]: https://github.com/itsmrwave/mpesa-gem/releases
[repositories]: https://github.com/itsmrwave/mpesa-gem
[rubydoc-mpesa]: http://rubydoc.info/gems/mpesa
[rubygems-mpesa]: http://rubygems.org/gems/mpesa


CHANGELOG.md
Changelog
=========

v0.1.0
------

* Add C2B process checkout request and response handling.

v0.0.0
------

Initial release, kind of a proof of concept ... by v1.0.0 we should have a
version ready for deployment in production environment and fairly tested.

* Initialise project.
* Add basic `MPesa::C2B::Client`.


spec/mpesa_spec.rb
require 'spec_helper'

describe MPesa do
  it 'has a version number' do
    expect(MPesa::VERSION).not_to be nil
  end
end


spec/spec_helper.rb
# The Coveralls.wear! must occur before any of our application code is
# required, so should be at the very top
require 'coveralls'
Coveralls.wear!

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

require 'faker'
require 'mpesa'
require 'webmock/rspec'

WebMock.disable_net_connect!(allow_localhost: true)

RSpec.configure do |config|
  config.expect_with :rspec do |c|
    # Disable should syntax
    c.syntax = :expect
  end
end


spec/c2b/client_spec.rb
require 'spec_helper'

module MPesa
  module C2B
    describe Client do
      describe '.VERSION' do
        it 'has the specification version defined' do
          expect(MPesa::C2B::VERSION).to eq '2.0'
        end
      end

      describe '.new' do
        it 'returns client instance' do
          client = MPesa::C2B::Client.new(
            merchant_id: Faker::Lorem.characters(16),
            passkey: Faker::Lorem.characters(8),
            callback_url: Faker::Internet.url('example.com'),
            callback_method: Faker::Lorem.characters(4)
          )
          expect(client).to be_an_instance_of(MPesa::C2B::Client)
        end
      end

      describe '#checkout' do
        let(:checkout_request) do
          stub_request(:post, 'https://safaricom.co.ke/mpesa_online/lnmo_checkout_server.php?wsdl')
            .with(headers: { 'Soapaction' => '"processCheckOut"' })
        end

        let(:checkout_response_success) do
          file_path = File.join(MPesa.gem_root, 'spec', 'support', 'files', 'checkout-response-success.xml')
          file = File.new file_path.to_s
          file.read
        end

        let(:checkout_response_failure) do
          file_path = File.join(MPesa.gem_root, 'spec', 'support', 'files', 'checkout-response-failure.xml')
          file = File.new file_path.to_s
          file.read
        end

        let(:transaction_details) do
          {
            transaction_id: 'TRANS-123/05',
            amount: 12_345.56,
            mobile_number: '254711222333',
            reference_id: '1112254500',
            data: { example_1: 'data 1', example_2: 'data 2', example_3: 'data 3' }
          }
        end

        before :context do
          @client = MPesa::C2B::Client.new(
            merchant_id: '123-456',
            passkey: 'sekret',
            callback_url: 'http://example.com/payments/mpesa',
            callback_method: 'POST'
          )
        end

        context 'when response is success' do
          it 'initiates a payment request' do
            checkout_request.to_return(status: 200, body: checkout_response_success)
            response = @client.checkout transaction_details

            expect(response.code).to eq '00'
            expect(response.failure?).to be false
            expect(response.message).to eq 'Example success message here'
            expect(response.success?).to be true
            expect(response.transaction_id).to eq 'cce3d32e0159c1e62a9ec45b67676200'
          end
        end

        context 'when response is failure' do
          it 'initiates a payment request' do
            checkout_request.to_return(status: 200, body: checkout_response_failure)
            response = @client.checkout transaction_details

            expect(response.code).to eq '01'
            expect(response.failure?).to be true
            expect(response.message).to eq 'Example failure message here'
            expect(response.success?).to be false
            expect(response.transaction_id).to eq 'cce3d32e0159c1e62a9ec45b67676200'
          end
        end
      end
    end
  end
end


lib/mpesa/c2b.rb
require 'mpesa/c2b/client'
require 'mpesa/c2b/response'

module MPesa
  # The overall scope of the C2B Web service is to provide primitives for
  # application developers to handle checkout process in a simple way. The API
  # allows merchant initiate C2B (paybill via web) transactions.
  #
  # ### How It Works
  #
  #     +-------------+                                       +---------------+
  #     |             |                                       |               |
  #     |             |>--(A)------- Payment Request -------->|               |
  #     |             |                                       |               |
  #     |             |<-------- Payment API Response -------<|               |
  #     |             |                                       |               |
  #     |        (B) -|                                       |               |
  #     |             |                                       |               |
  #     |  MERCHANT   |>--(C)--- Confirm Payment Request ---->|      SAG      |
  #     |             |                                       |               |
  #     |             |<-------- Confirm API Response -------<|               |
  #     |             |                                       |               |
  #     |             |                                       |- (D)          |
  #     |             |                                       |               |
  #     |             |<--(H)-- Payment Response Callback ---<|               |
  #     |             |                                       |               |
  #     |             |>--(I)--- Callback Result Response --->|               |
  #     |             |                                       |               |
  #     |             |                                       |               |
  #     +-------------+                                       +---------------+
  #
  # * **A** - The merchant submits authentication details, transaction details,
  #   callback url and callback method. Merchant receives instant feedback with
  #   validity status of their payment request.
  # * **B** & **C** - Customer is shown a payment request message and a
  #   confirmation payment request is initiated after. Merchant receives instant
  #   feedback with validity status of their confirm payment request.
  # * **D** - C2B API handles customer validation and authentication via USSD
  #   push.
  #     * **H** - If the customer validation fails or declines the transaction,
  #       the API makes a callback to merchant.
  #     * **I** - Otherwise the transaction is processed and status is reported
  #       back through a callback.
  #
  # This process flow makes the following assumptions:
  #
  # * The service is an on demand service where the Merchant always initiates
  #   the request when there is a payment to be settled.
  # * The merchant already has an account created by the SAG.
  # * The merchant has been provided with the URL of the endpoint.
  module C2B
    VERSION = '2.0'.freeze
  end
end


lib/mpesa/c2b/response.rb
module MPesa
  module C2B
    # Response handler for {MPesa::C2B::Client} requests.
    #
    # This class is used by {MPesa::C2B::Client} internally to process request
    # responses and is not intended to be used in isolation.
    class Response
      # Creates a new instance of {MPesa::C2B::Response}.
      #
      # @param response [Savon::Response] Savon Response object.
      def initialize(response)
        @response = response
      end

      # Gets the response code
      #
      # @return [String]
      def code
        body[:return_code].to_s
      end

      # Check if the request made initially resulted in a failure.
      #
      # @return [Boolean]
      def failure?
        !success?
      end

      # Gets the response message.
      #
      # @return [String]
      def message
        body[:cust_msg].to_s
      end

      # Check if the request made initially was successful.
      #
      # @return [Boolean]
      def success?
        case body[:description].to_s
        when 'Success' then true
        when 'Failure' then false
        end
      end

      # Gets the request-reponse transaction id
      #
      # @return [String]
      def transaction_id
        body[:trx_id].to_s
      end

      private

      # Get the body of the response.
      #
      # @return [Hash]
      def body
        @response.body[:process_check_out_response]
      end
    end
  end
end


lib/mpesa/c2b/client.rb
module MPesa
  module C2B
    # Client interface for {MPesa::C2B}
    class Client
      # Creates a new instance of {MPesa::C2B::Client}.
      #
      # Example:
      #
      #     client = MPesa::C2B::Client.new(
      #       merchant_id: '123-456',
      #       passkey: 'sekret',
      #       callback_url: 'http://example.com/payments/mpesa',
      #       callback_method: 'POST'
      #     )
      #
      # @param merchant_id: [String] The merchant's unique transaction
      #     identifier.
      #
      # @param passkey: [String] Merchant's passwork key provided by Safaricom.
      #
      # @param callback_url: [String] The merchant's payment notification
      #     endpoint URL. Transaction status and the request results will be
      #     sent to this endpoint.
      #
      # @param callback_method: [String] The callback method used to send
      #     parameters to the perchant's payment notification endpoint defined
      #     by `callback_url`.
      def initialize(merchant_id:, passkey:, callback_url:, callback_method:)
        @merchant_id     = merchant_id
        @passkey         = passkey
        @callback_url    = callback_url
        @callback_method = callback_method

        @client = Savon.client do
          soap_version 1
          env_namespace :soapenv
          wsdl File.join(::MPesa.gem_root, 'lib', 'mpesa', 'c2b', 'wsdl.xml').to_s
        end
      end

      # Request the SAG to start the C2B process to complete the M-Pesa payment.
      #
      # Example:
      #
      #     transaction_details = {
      #       transaction_id: 'TRANS-123/05',
      #       amount: 12_345.56,
      #       mobile_number: '254711222333',
      #       reference_id: '1112254500',
      #       data: { example_1: 'data 1', example_2: 'data 2', example_3: 'data 3' }
      #     }
      #
      #     response = client.checkout transaction_details
      #
      #     puts response.code # => '00'
      #     puts response.success # => true
      #     puts response.transaction_id # => 'cce3d32e0159c1e62a9ec45b67676200'
      #
      # @param amount: [Float] value of the transaction.
      #
      # @param data: [Hash] merchant data that we want to pass along with the
      #   request. It may be anything that may help with transaction processing
      #   later on.
      #
      # @param mobile_number: [String] the customer's mobile phone number.
      #
      # @param reference_id [String] identifier of the product or service being
      #     offered.
      #
      # @param transaction_id [String] unique transaction identifier (respective
      #     to the merchant).
      #
      # @return [MPesa::C2B::Response]
      def checkout(transaction_id:, amount:, mobile_number:, reference_id:, data: {})
        headers  = payment_request_headers
        body     = payment_request_body(transaction_id, amount, mobile_number, reference_id, data)
        response = call __method__, headers, body
        Response.new response
      end

      private

      # Constants
      OPERATION_MAPPING = { checkout: :process_check_out }.freeze

      # Call SOAP operation.
      #
      # @return [Savon::Response]
      def call(method, headers, body)
        operation = OPERATION_MAPPING[method].to_sym
        @client.call(operation) do |locals|
          locals.soap_header headers
          locals.message body
        end
      end

      # Generate password for use with authenticating with SAG.
      #
      # @return [String]
      def password(timestamp)
        string = [@merchant_id, @passkey, timestamp].join(' ')
        hashed_string = Digest::SHA256.hexdigest string
        Base64.encode64(hashed_string).upcase
      end

      # Build SOAP header parameters for the checkout request.
      #
      # Savon 2 doesn't generate header namespaces correctly so we have to sort
      # out that special case appopriately. See: https://github.com/itsmrwave
      # /mpesa-gem/issues/5
      #
      # @return [Hash]
      def payment_request_headers
        { 'tns:CheckOutHeader' => request_headers }
      end

      # Build SOAP body parameters for the checkout request.
      #
      # @return [Hash]
      def payment_request_body(transaction_id, amount, mobile_number, reference_id, data = {})
        body = {
          'MERCHANT_TRANSACTION_ID' => transaction_id,
          'REFERENCE_ID' => reference_id.to_s,
          'AMOUNT' => amount.to_s,
          'MSISDN' => mobile_number.to_s,
          'CALL_BACK_URL' => @callback_url,
          'CALL_BACK_METHOD' => @callback_method
        }
        body['ENC_PARAMS'] = Base64.encode64(data.to_s) unless data.nil?
        body
      end

      # Generate base headers used as authentication and validation parameters.
      #
      # @return [Hash]
      def request_headers
        timestamp = Time.now.to_i.to_s
        {
          'MERCHANT_ID' => @merchant_id,
          'PASSWORD' => password(timestamp),
          'TIMESTAMP' => timestamp
        }
      end
    end
  end
end
